File: elixir_leetcode/lib/day06/max_sub_array.ex
defmodule MaxSubArray do
  @spec max_sub_array(nums :: [integer]) :: integer
  def max_sub_array(nums) do
    exec2(nums)
  end


  # -1 * abs(hd(nums)) 只是要取一個初始值而已，這個值永遠不能比一個數字大
  def exec2(nums) do
    Enum.scan(nums, -1 * abs(hd(nums)), &max(&1 + &2, &1))
    |> Enum.max()
  end


  def exec(nums, max_sum \\ nil, sofar_sum \\ nil)
  def exec([num | tail], nil, nil) do
    exec(tail, num, num)
  end
  def exec([], max_sum, _sofar_sum) do
    max_sum
  end
  def exec([num | tail], max_sum, sofar_sum) do
    next_sofar_sum = max(sofar_sum + num, num)
    exec(tail, max(next_sofar_sum, max_sum), next_sofar_sum)
  end
end

File: ts_leetcode/day06/max_sub_array.ts


export function maxSubArray(nums: number[]): number {
  let maxSum = 0

  for (let from = 0; from < nums.length; from++) {
    for (let to = 0; to < nums.length; to++) {
      const sum = nums.slice(from, to + 1).reduce((s, num) => s + num, 0);
      
      
      maxSum = Math.max(maxSum, sum)
    }
  }

  return maxSum
}

export function maxSubArrayKadane(nums: number[]): number{
  let maxSum: number = nums[0];
  let sofarSum: number = nums[0];

  for (let i = 1; i < nums.length; i++) {
    const num = nums[i];
    sofarSum = Math.max(sofarSum + num, num);
    maxSum = Math.max(maxSum, sofarSum);
    
  }

  return maxSum
}
